### DISCUSSION TITLE
Scan RESTEasy resources in a library outside deployment files

### DISCUSSION THREAD

**Discussion by tituslup** [resteasy/resteasy#4141] [Q&A] [answered]
Hi,

I'm trying to add REST API to an existing wildfly (30) application using RESTEasy (6.2.5).
So far the only success I had was with RESTEasy resources (tagged with @Path, @GET etc) that are included in ear or war files either as classes or libraries. Application and resource detection works, endpoints work.

The problem I'm facing is that I want to add annotations in existing classes part of a library that is not bundled in my war/ear files. That library is shared with all deployments using a global wildfly module. It doesn't look like there is any resource detection on the external library. Applications are functional and able to import from that library, its just that resource detection is not working and as result no REST API.

```xml
<module xmlns="urn:jboss:module:1.3" name="com.prod.api">
  <resources>
    <resource-root path="prod_api.jar"/>
  </resources>
  <dependencies>
      <module name="jakarta.ws.rs.api" services="import"/>
      <module name="org.jboss.resteasy.resteasy-core" services="import"/>
      <module name="org.hibernate.validator" services="import"/>
      <module name="javax.ejb.api" export="true"/>
      <module name="javax.api" export="true"/>
  </dependencies>
</module>
```

Here is a **sample resource** class I know works if I place it in my deployment files, but does not work when I place it in the external library.

```java
package testpkg;

import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.QueryParam;
import jakarta.ws.rs.Consumes;
import jakarta.ws.rs.Produces;


@Path("/resources")
@Consumes({"application/json"})
@Produces({"application/json"})
public class APIRestResource {

    @GET
    @Path("/refs")
    public String findRefs(@QueryParam("filter") String filter) {
        return "APIRestResource";
    }
}
```

Can you point me in the right direction and be able to scan that external library?

Thanks,
Titus

**Comment by jamezp** [comment]
How are you adding your module dependency to your deployment?

**Comment by tituslup** [comment]
First its listed as a dependency in another, quite large, global module:
```xml
<module xmlns="urn:jboss:module:1.3" name="com.arc">
    <resources>
        ...
    </resources>
    <dependencies>
        <module name="javax.api" export="true"/>
        <module name="javax.persistence.api" export="true"/>
        <module name="javax.enterprise.concurrent.api" export="true"/>
        <module name="javax.servlet.api" export="true"/>
        <module name="com.prod.api" export="true"/>
        ...
    </dependencies>
</module>
```

Then this larger module is listed under each application jboss-deployment-structure.xml:
```xml
<jboss-deployment-structure>
    <sub-deployment name="common.jar">
        <dependencies>
            <module name="jakarta.ws.rs.api" services="none"/>
            <module name="org.jboss.resteasy.resteasy-core" services="import"/>
            <module name="org.hibernate.validator" services="import"/>
            <!-- Tried here directly at one point    <module name="com.prod.api" export="true"/>  -->
        </dependencies>
    </sub-deployment>
</jboss-deployment-structure>
```

**Reply by jamezp** [reply]
Is the `common.jar` in the deployment?

I"m not sure if I'll have time today, but I can try to create a reproducer to see resources from a module are registered. I want to think they are, but you _may_ need to add the providers in a `META-INF/services/jakarta.ws.rs.ext.Providers` file. That is very RESTEasy specific though and something I'd like to deprecate the usage of.

**Reply by tituslup** [reply]
Yes its part of the deployment ear file in this case

Tried a simplified example, still using the same "com.prod.api" global module, added as dependency directly in dummy application with another jboss-deployment-structure.xml:
```xml
<jboss-deployment-structure>
    <deployment name="restapi.war">
        <dependencies>
            <module name="jakarta.ws.rs.api" services="import"/>
            <module name="com.prod.api" services="import"/>
        </dependencies>
    </deployment>
</jboss-deployment-structure>
```

Then deployed this new war on the system:
```
2024-04-19 15:48:18,163 INFO Thread-21 [org.jboss.as.server.deployment.AbstractDeploymentUnitService] [start] WFLYSRV0027: Starting deployment of "restapi.war" (runtime-name: "restapi.war")
2024-04-19 15:48:18,241 INFO Thread-337 [org.jboss.resteasy.core.ResteasyDeploymentImpl] [processApplication] RESTEASY002225: Deploying jakarta.ws.rs.core.Application: class com.restapi.TestApplication
2024-04-19 15:48:18,244 INFO Thread-337 [org.wildfly.extension.undertow.Host] [registerDeployment] WFLYUT0021: Registered web context: '/restapi' for server 'default-server'
2024-04-19 15:48:18,261 INFO Thread-29 [org.jboss.as.server.deployment.DeploymentHandlerUtil$1$2] [handleResult] WFLYSRV0010: Deployed "restapi.war" (runtime-name : "restapi.war")
```

Resources found inside the war, their endpoints work.
The one from prod_api.jar does not. Importing a class from there works.
Maybe I need a dummy jar with a resource to exclude my library sa the problem.

**Reply by tituslup** [reply]
So I've tried to exclude all of the existing code (I hope I did) and do a dummy example.

This is the jar with only the Resource provided at the begining.
```
# pwd
    /opt/wildfly30/modules/com/prod/api/main
# ls
    module.xml  prod_api.jar
# unzip prod_api.jar
    Archive:  prod_api.jar
       creating: META-INF/
      inflating: META-INF/MANIFEST.MF
       creating: testpkg/
      inflating: testpkg/APIRestResource.class
```

I hooked this up to a new wildfly30 instance with mostly default config in standalone-full.xml (ports and public interface are different).
Code from restapi.war does an import from testpkg just to confirm dependency works. It does.
Resource from restapi.war is functional, the one from prod_api.jar is not.

Maybe you can point out something I'm doing wrong.
I'll have a look over jakarta.ws.rs.ext.Providers as you pointed earlier.

Thanks!

**Reply by jamezp** [reply]
I know there is some scanning that only happens on WAR's. I don't think resources in a `*.jar` are scanned, but I'd definitely need to double check that. Most of the scanning itself happens in WildFly and setups RESTEasy accordingly.

**Comment by tituslup** [comment, marked as answer]
I have decided to include the jar in my deployment files. Scan is obviously working fine this way and it gets me quicker to the end goal (adding a REST API interface).

 jboss-deployment-structure.xml:
```xml
<jboss-deployment-structure>
    <deployment name="restapi.war">
        <dependencies>
            <module name="jakarta.ws.rs.api" services="import"/>
        </dependencies>
        <resources>
            <resource-root path="prod_api.jar" />
        </resources>
    </deployment>
</jboss-deployment-structure>
```

**Reply by jamezp** [reply]
You shouldn't need to add the `jakarta.ws.rs.api` module as it will automatically be added for you. If the `prod_api.jar` is also in `restapi.war/WEB-INF/lib` directory.

**Reply by tituslup** [reply]
Thanks for the tips, will test these out.
